.adr/0017-missed-chain-events.md

```diff
@@ -0,0 +1,127 @@
+
+### 1. Put `Channel` data protection logic in `Channel` class
+
+Add logic to the `Channel` class to protect its data from stale updates rather than relying on an external component (`chainservice`) to only feed it sanitized data. This seems less error-prone and it could be useful to add a `Channel.LastUpdated` struct in case a user wants to know when the channel was last changed. The `chainservice` will still be expected to feed chain events in order, but the added `Channel` logic will act as another layer of protection against faulty data updates.
+
+The `Channel` class should keep track of the `TxIndex` associated with each chain event in case there are multiple events related to a single channel within the same block. Also, an `ethereum.FilterQuery` cannot specify a `TxIndex` so it will be more computationally efficient to check the `TxIndex` in the `Channel` class rather than in the `chainservice` logic.
+
+```go
+type Channel struct {
+	state.FixedPart
+	Id      types.Destination
+	MyIndex uint
+
+	OnChain  OnChainData
+	OffChain OffChainData
+
+  LastChainUpdate ChainUpdateData
+}
+
+type ChainUpdateData struct {
+        BlockNum uint64
+        TxIndex uint64
+}
+
+func (c *Channel) UpdateWithChainEvent(event chainservice.Event) (*Channel, error) {
+	if event.BlockNum() > c.LastChainUpdate.BlockNum && event.TxIndex() > c.LastChainUpdate.TxIndex {
+		// Process event
+		...
+
+		// Update Channel.LastChainUpdate
+		c.LastChainUpdate.BlockNum = event.BlockNum()
+		c.LastChainUpdate.TxIndex  = event.TxIndex()
+		return c, nil
+	}
+	return nil, fmt.Errorf("chain event older than channels last update")
+}
+```
+
+### 2. Updating `lastBlockNum`: `engine` periodically reads from `chainservice`
+
+Instead of the `chainservice` alerting the `engine`, the `engine` can request the block number from the `chainservice`. This means the `lastBlockNumSeen` wouldn't get updated every block, but that's not a major concern since the node
+
+The benefit of this approach is simplicity and limited additional strain on the `engine`/`store`. Instead of adding a new chan or event, a `chainservice.GetLastConfirmedBlockNum()` function can be added and called by the `engine` periodically. This also means the node wouldn't write to the `store` every block but every x seconds (and when the node closes). The `engine` run loop can be updated to this:
+
+```go
+	blockTicker := time.NewTicker(15 * time.Second)
+
+	select {
+		...
+		case chainEvent := <-e.fromChain:
+			res, err = e.handleChainEvent(chainEvent)
+		case <-blockTicker.C:
+			blockNum := e.chain.GetLastConfirmedBlockNum()
+			err = e.store.SetLastBlockNumSeen(blockNum)
+	}
+```
+
+## Future considerations
+
+1. Is there any situation where an older chain event SHOULD be allowed to update a `Channel` (i.e. chain re-org)? We could allow a way to forcibly process an old chain event, or we could assume that the node is fully protected against any situation that would necessitate that action.
```

Comment:
> We could allow a way to forcibly process an old chain event,
I'm not sure if that would be enough to handle chain re-orgs, since we'd also have to handle a case where an event happens and then after the chain-org the event didn't happen. So we'd need a mechanism to detect an event no longer exists and "undo" it.

I think there may also be an inherit risk to chain re-orgs that we can really only solve by waiting? That's because chain events can trigger nitro nodes to exchanged signed states, and once these are sent out we can't really undo them. For example if Bob signs a state with outcome `[A:5,B:5]`  based on-chain channel balance of `10`, he's at risk if a chain re-org occurs and the on-chain channel balance  is actually `5` now.

Reply:
Yes that makes sense about having to "undo" off-chain state changes. I am happy to remove the "Future considerations" section to make this adr cleaner. Would that be preferable or do you think this adr needs more thought and detail about how the node should handle chain re-orgs that defeat the `REQUIRED_BLOCK_CONFIRMATIONS` limit?

Reply:
I'm happy to remove this section (or even leave it) but I think further thought about chain re-orgs should probably be handled outside of this PR/ADR (due to the complexity).

Reply:
Removed that section: https://github.com/statechannels/go-nitro/pull/1624/commits/f69ab0eb34a0dcbc8c65c70ac2edfd051a682047